packages/frontend/src/main/components/stream/favorites/StreamFavoriteBtn.vue

```diff
@@ -0,0 +1,121 @@
+<template>
+  <v-btn v-if="user && canFavorite" icon color="red darken-3" @click="onFavoriteClick">
+    <v-icon>
+      {{ isFavorited ? 'mdi-heart' : 'mdi-heart-outline' }}
+    </v-icon>
+  </v-btn>
+</template>
+
+<script>
+import gql from 'graphql-tag'
+import { canBeFavorited } from '@/helpers/streamHelpers'
+import { UserFavoriteStreamsQuery } from '@/graphql/user'
+import { COMMON_STREAM_FIELDS } from '@/graphql/streams'
+
+export default {
+  name: 'StreamFavoriteBtn',
+  props: {
+    stream: { type: Object, required: true },
+    user: { type: Object, required: true }
+  },
+  computed: {
+    isFavorited() {
+      return !!this.stream.favoritedDate
+    },
+    canFavorite() {
+      return canBeFavorited(this.stream)
+    }
+  },
+  methods: {
+    async onFavoriteClick(e) {
+      e.preventDefault() // Preventing click on the parent link
+
+      const newIsFavorited = !this.isFavorited
+      const { id, favoritesCount } = this.stream
+
+      // Pre-generate optimistic results
+      const newFavoritedDate = newIsFavorited ? new Date().toISOString() : null
+      const newFavoritesCount = favoritesCount + (newIsFavorited ? 1 : -1)
+
+      // Toggle favorited status
+      await this.$apollo.mutate({
+        mutation: gql`
+          mutation ($sid: String!, $favorited: Boolean!) {
+            streamFavorite(streamId: $sid, favorited: $favorited) {
+              id
+              favoritedDate
+              favoritesCount
+            }
+          }
+        `,
+        variables: {
+          sid: this.stream.id,
+          favorited: newIsFavorited
+        },
+        optimisticResponse: {
+          __typename: 'Mutation',
+          streamFavorite: {
+            __typename: 'Stream',
+            id,
+            favoritedDate: newFavoritedDate,
+            favoritesCount: newFavoritesCount
+          }
+        },
+        update: (cache, { data: { streamFavorite } }) => {
+          const { id, favoritedDate } = streamFavorite || {}
```

Comment:
This part's cool - using the update function I'm updating the results of the "favorite streams" query, so that favorite streams are added/removed without needing to refetch from server

Reply:
I'm also being very careful to not mutate existing cache objects, in case you're wondering why I don't replace property values on old objects directly